ogr/ogrsf_frmts/miramon/mm_wrlayr.c

```diff
@@ -0,0 +1,7038 @@
+    {
+        if (MM_WriteNRecordsMMBD_XPFile(pMMAdmDB))
+            return 1;
+
+        // Flushing all to be flushed
+        pMMAdmDB->FlushRecList.SizeOfBlockToBeSaved = 0;
+        if (MMAppendBlockToBuffer(&pMMAdmDB->FlushRecList))
+            return 1;
+
+        pMMAdmDB->pMMBDXP->pfDataBase = pMMAdmDB->pFExtDBF;
+
+        if (MM_ChangeDBFWidthField(
+                pMMAdmDB->pMMBDXP, nIField, nNewWidth,
+                pMMAdmDB->pMMBDXP->pField[nIField].DecimalsIfFloat,
+                (MM_BYTE)MM_NOU_N_DECIMALS_NO_APLICA))
+            return 1;
+
+        // The record on course also has to change its size.
+        if ((GUInt64)pMMAdmDB->pMMBDXP->BytesPerRecord + 1 >=
+            pMMAdmDB->nNumRecordOnCourse)
+        {
+            if (nullptr == (pMMAdmDB->szRecordOnCourse = realloc_function(
+                                pMMAdmDB->szRecordOnCourse,
+                                (size_t)pMMAdmDB->pMMBDXP->BytesPerRecord + 1)))
+            {
+                MMCPLError(CE_Failure, CPLE_OutOfMemory,
+                           "Memory error in MiraMon "
+                           "driver (MMTestAndFixValueToRecordDBXP())");
+                return 1;
+            }
+        }
+
+        // File has changed it's size, so it has to be updated
+        // at the Flush tool
+        fseek_function(pMMAdmDB->pFExtDBF, 0, SEEK_END);
+        pMMAdmDB->FlushRecList.OffsetWhereToFlush =
+            ftell_function(pMMAdmDB->pFExtDBF);
+    }
+    return 0;
+}
+
+int MMWriteValueToRecordDBXP(struct MiraMonVectLayerInfo *hMiraMonLayer,
+                             char *registre, const struct MM_FIELD *camp,
+                             const void *valor, MM_BOOLEAN is_64)
+{
+    if (!hMiraMonLayer)
+        return 1;
+
+    if (!camp)
+        return 0;
+
+    if (MMResizeStringToOperateIfNeeded(hMiraMonLayer,
+                                        camp->BytesPerField + 10))
+        return 1;
+
+    if (!valor)
+        *hMiraMonLayer->szStringToOperate = '\0';
+    else
+    {
+        if (camp->FieldType == 'N')
+        {
+            if (!is_64)
+            {
+                sprintf(hMiraMonLayer->szStringToOperate, "%*.*f",
+                        camp->BytesPerField, camp->DecimalsIfFloat,
+                        *(const double *)valor);
+            }
+            else
+            {
+                sprintf(hMiraMonLayer->szStringToOperate, "%*lld",
+                        camp->BytesPerField, *(const GInt64 *)valor);
+            }
+        }
+        else
+        {
+            sprintf(hMiraMonLayer->szStringToOperate, "%-*s",
+                    camp->BytesPerField, (const char *)valor);
+        }
+    }
+
+    memcpy(registre + camp->AcumulatedBytes, hMiraMonLayer->szStringToOperate,
+           camp->BytesPerField);
+    return 0;
+}
+
+// Gets the n-th value of the format (number_of_values:val1,val2,...,valN)
+char *MMGetNFieldValue(const char *pszStringList, GUInt32 nIRecord)
```

Comment:
I do feel you could just remove that function.  Why not directly using the values[i] from the poFeature->GetFieldAsInteger/Integer64/DoubleList(iField, &nCount) and formatting them as a string with CPLSPrintf() in ogrmiramonlayer.cpp ? That would avoid that error-prone function

Reply:
I don't know why I didn't do that at the beginning.

So, there is three cases:
Integer:
CPLSPrintf("%d", panValues[nIRecord]),
Integer64
CPLSPrintf("%" CPL_FRMT_GB_WITHOUT_PREFIX, panValues[nIRecord]),
Double
CPLSPrintf("%f", panValues[nIRecord]),

And how many decimals are going to be written?

Reply:
I think I did to the double case, to mantain the same number of decimals than the origin

Reply:
> And how many decimals are going to be written?

same as printf("%f"), so according to https://man7.org/linux/man-pages/man3/printf.3.html  6 decimals.  You can increase it for example with ``"%.8f"`` if needed

Reply:
And how to know how many decimals have the source number panValues[nIRecord]?
Here was my problem, I think

Reply:
> And how to know how many decimals have the source number panValues[nIRecord]?

you can't know. A IEEE754-double precision may require up to 18 significant digits to be faithfully represented as decimal without losing any bit. That said, if unsure, you use "%f" and that should be fine for most use cases

Reply:
But it's a pitty that something like that: 8400.900000001 at the end of conversion is gonna be 8400.900000
Perhaps this decimal is important. On the other way something like 84683.8 could become 84683.8000000001 if there are too many decimals. There is any way to indicate how many decimals you want in a translation?
I read something about that, but I don't know if could apply here?

Reply:
Just use "%f" for now

Reply:
At some time my boss will ask me for that. Why only 6 decimals? And at that time the conversation will arise again :)